src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -518,5 +649,106 @@ internal JsonArrayPool(ArrayPool<T> arrayPool)
 
             public void Return(T[] array) => this.arrayPool.Return(array);
         }
+
+        private class JsonProgressClientConverter : JsonConverter
+        {
+            private readonly JsonMessageFormatter formatter;
+
+            public JsonProgressClientConverter(JsonMessageFormatter formatter)
+            {
+                this.formatter = formatter ?? throw new ArgumentNullException(nameof(formatter));
+            }
+
+            public override bool CanConvert(Type objectType) => MessageFormatterHelper.FindIProgressOfT(objectType) != null;
+
+            public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+            {
+                throw new NotSupportedException();
+            }
+
+            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+            {
+                // if the requestId is empty it means the Progress object comes from a response or a notification
+                if (this.formatter.requestId == null)
+                {
+                    throw new NotSupportedException("IProgress<T> objects should not be part of any response or notification.");
+                }
+
+                lock (this.formatter.progressMap)
+                {
+                    long progressId = this.formatter.nextProgressId++;
+                    this.formatter.requestProgressMap.Add(this.formatter.requestId.Value, progressId);
+
+                    this.formatter.progressMap.Add(progressId, new ProgressParamInformation(value));
+
+                    writer.WriteValue(progressId);
+                }
+            }
+        }
+
+        private class ProgressParamInformation
+        {
+            public Type iProgressOfTType;
+            public Type valueType;
+            public MethodInfo reportMethod;
+            public object progressObject;
+
+            public ProgressParamInformation(object progressObject)
+            {
+                this.iProgressOfTType = MessageFormatterHelper.FindIProgressOfT(progressObject.GetType());
+                this.valueType = this.iProgressOfTType.GenericTypeArguments[0];
```

Comment:
This `iProgressOfTType` is never used outside this ctor, so it needn't be a field. It can just be a local variable.
That's the only one I checked. Are all the others needed as fields?